### PR TITLE
Create snooze task endpoint 

### LINKED ISSUE #7368: Create snooze task endpoint
An endpoint should be created that allows tasks to be "snoozed."  This is currently the "Remind me later" option, though I think we could update the language to match that used in notes.

Currently these options are stored in `woocommerce_task_list_remind_me_later_tasks`.  If we want to retain backwards compatibility, the data can remain in this option without affecting existing stores.

We will want to filter the task list created in https://github.com/woocommerce/woocommerce-admin/issues/7365 to include `isSnoozed` and possibly `snoozedUntil` in each of the tasks.

**Endpoint**

`POST` - `/wc-admin/onboarding/tasks/<task_id>/snooze`

**Params**

While this is currently a single time, it may be worth adding a single param for duration of the snooze.

* `duration` - integer

**Response**

This endpoint can simply return a response message if successful or contain errors if there was a problem snoozing the task.

### PR DESCRIPTION
**Pull request by joelclimbsthings**
Fixes #7368

**Add Endpoint**:
`POST - /wc-admin/onboarding/tasks/<task_id>/snooze`

**Parameters:**
- `duration`: Optional, currently supports `day|hour|week`, and defaults to `day`
- `task_list_id`: Optional, constrains search to specific task list

**Response**
Currently returns an error if the task does not exist, or the modified task definition. 

**Notes:**
- Also adding `isSnoozed` and `snoozedUntil` properties to the task definitions returned by `GET /wc-admin/onboarding/tasks`
- ~~Haven't implemented tracks yet (I'll be able to get this done early Monday)~~

### Detailed test instructions:

-   Checkout branch
-   Use a tools such as PostMan to send a request to `POST - /wc-admin/onboarding/tasks/invalid-task/snooze`
- Confirm that an error response is returned
- Now take a valid task ID and do the same request to `POST - /wc-admin/onboarding/tasks/<task_id>/snooze`
- Ensure that it returns the modified task definition with valid `isSnoozed` and `snoozedUntil` properties.
- Make a request to `GET - /wc-admin/onboarding/tasks` and check the definition of that task to ensure those properties are present and valid.

### CODE REVIEW

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:45]: We have constants like `DAY_IN_SECONDS` that may be useful here instead of a new variable.

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:927]: I think this could be added via filter to keep the logic down in this method.  See https://github.com/woocommerce/woocommerce-admin/pull/7538

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:961]: Nice function!  In https://github.com/woocommerce/woocommerce-admin/pull/7538 I moved the tasks to the `OnboardingTasks` data store.  This method might also be more appropriate there.

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:963]: This will only get the first task list, which could create confusion.  Since we can have multiple task lists, I think we'll need to loop over all of them.

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:930]: Should this check to see if the snooze time has already passed?

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:987]: I think we may need a property `isSnoozeable` in here (previously `allowRemindMeLater`) to determine whether a task is allowed to be snoozed or not.

**Review by joshuatf — COMMENTED** [review]
Thanks for getting this endpoint up and the extra utility functions!  Left a few comments in the code.

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: So far I decided to keep the mapping since I find it useful in accessing this in this class, but I'm now utilizing these constants. Thanks for the heads up 👍🏻

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: Good call, done in 22297e280f1b810ab4484050a54119574913e499 👍🏻

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: I like it, done.

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: Good call, fixed 👍🏻

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php:799]: Right, this is now implemented.

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: Perhaps, I'd assumed this option was already being managed via a CRON operation, but I'm glad to address this. I'm not sure what action we'd want to take in this case?

**Comment by joelclimbsthings** [comment]
Thanks for the review @joshuatf ! I believe I've addressed your comments with the only exception being if some action should take place if the `snoozeUntil` timestamp has already passed. Any thoughts on what should occur in this case? 

Quick notes:
- Also added a optional param `task_list_id` to the `/snooze` endpoint as an optimization, and to avoid potential id conflicts across lists. 
- I noticed that the current logic supports undoing a snooze, although this wasn't mentioned in the original issue. Glad to add this; perhaps in a follow-up PR?

File: src/API/OnboardingTasks.php
Comment (joshuatf) [reply src/API/OnboardingTasks.php]: I don't think this needs to be managed by cron.  If we store a timestamp with the date a task is snoozed until, we can compare that we when retrieve timestamps and check wether or not they're still snoozed:

```php
$task[ 'isSnoozed' ] = $snoozed_until > now();
```

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:745]: Thanks for using your new util here!  Nitpicky, but this variable sounds like a verb.  Maybe just `$task`?

```suggestion
		$task = OnboardingTasksFeature::get_task_by_id( $id );
```

File: src/Features/OnboardingTasks.php
Comment (joshuatf) [inline src/Features/OnboardingTasks.php:803]: This bool can be dynamically created.   I haven't tested this, but something along the lines of:

```suggestion
					$task_lists[ $task_list_key ]['tasks'][ $task_key ]['isSnoozed']    = $snoozed_tasks[ $task['id'] ] > time();
```

**Review by joshuatf — COMMENTED** [review]
Thanks for the changes, @joelclimbsthings!  Left a comment on one more item to address before merge.

>Also added a optional param task_list_id to the /snooze endpoint as an optimization, and to avoid potential id conflicts across lists.

Sounds good!  I think the list of tasks and retrieval could benefit from a better API, but maybe we can address this at the end of the cycle if time allows or during cooldown.  In the meantime, this is a good change 👍 

>I noticed that the current logic supports undoing a snooze, although this wasn't mentioned in the original issue. Glad to add this; perhaps in a follow-up PR?

Ya, we can either keep it on this endpoint and allow a `duration` param of `none` or `null` to undo the snooze or we can create a new endpoint in a follow-up.  What do you think is best?

File: src/Features/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/Features/OnboardingTasks.php]: Ah, understand what you mean here now. Done in https://github.com/woocommerce/woocommerce-admin/pull/7539/commits/1979e575c0431127490c2ded6dffc98872bd22dc.

**Comment by joelclimbsthings** [comment]
> Ya, we can either keep it on this endpoint and allow a duration param of none or null to undo the snooze or we can create a new endpoint in a follow-up. What do you think is best?

I'd say the latter option aligns more with what I would expect if consuming the API. Should be a quick follow-up, and I'm glad to create the issue and get that done today.

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:157]: This arg should be removed; we can use the `id` retrieved from the path.

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:801]:
```suggestion
		$task_id         = $request->get_param( 'id' );
```

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:816]: We can also add a status code here.

```suggestion
				__( 'Sorry, no snoozeable task with that ID was found.', 'woocommerce-admin' ),
				array(
					'status' => 404,
				)
```

File: src/API/OnboardingTasks.php
Comment (joshuatf) [inline src/API/OnboardingTasks.php:820]: Needs default param to avoid error on initial snooze:

```suggestion
		$snooze_option = get_option( 'woocommerce_task_list_remind_me_later_tasks', array() );
```

File: src/Features/OnboardingTasks.php
Comment (joshuatf) [inline src/Features/OnboardingTasks.php:753]: Not a blocker to getting this merged, especially since we may update the task API by the end of cycle, but just a quick note that this loops over the tasks twice.

This most likely will not have any meaningful impact on performance given how small the task list is, but I'd opt for a single loop where possible if we update this.

**Review by joshuatf — APPROVED** [review]
Looks good, Joel!  Encountered a few minor errors and left suggestions on fixes, but pre-approving pending those fixes.

Thanks for the work on this 💯

File: src/Features/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/Features/OnboardingTasks.php]: I don't believe this is really any different than the nested `foreach` loops we'd be using otherwise? The reduce only loops through the task lists (only 1 currently), and then the following `foreach` loop just loops through the relevant tasks. The benefit is that if `$task_list_id` is provided, you'll skip the first loop entirely and only loop through the tasks for the specified list. 

This depends a bit on the performance of `array_merge`; is that what you're referring to? Not too familiar with the inner-workings there.

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: Good catch 👍🏻

File: src/API/OnboardingTasks.php
Comment (joelclimbsthings) [reply src/API/OnboardingTasks.php]: Whoops, that must have snuck in when I rebased 🤭

File: src/Features/OnboardingTasks.php
Comment (joshuatf) [reply src/Features/OnboardingTasks.php]: Yes and no.  You're correct if we simply nest the foreach and let it run.  However, with the nested option we can return early or break the loop which has the ability to be more performant.

But again, this is a very nitpicky comment and I'm hoping we can create a better API around registering/getting tasks in the backend that would make this obsolete anyway.

Comment (joshuatf) [reply src/Features/OnboardingTasks.php]: Yes and no.  You're correct if we simply nest the foreach and let it run.  However, with the nested option we can return early or break the loop which has the ability to be more performant.

But again, this is a very nitpicky comment and I'm hoping we can create a better API around registering/getting tasks in the backend that would make this obsolete anyway.

Comment (joelclimbsthings) [reply src/Features/OnboardingTasks.php]: True, although the reduce is only looping through the list objects, and the subsequent `foreach` loop that goes through the tasks does return early as you mention. Hopefully all moot as we improve this, but I was just curious if I was missing something. 🙏

**Pull request merged**
